app/main.py

```diff
@@ -1,2 +1,103 @@
 class Dictionary:
-    pass
+    def __init__(
+            self, initial_capacity: int = 8,
+            load_factor: float = 0.75
+    ) -> None:
+        self.capacity = initial_capacity
+        self.size = 0
+        self.buckets = [None] * self.capacity
+        self.load_factor = load_factor
+
+    def __hash_key(self, key: any) -> int:
+        return hash(key) % self.capacity
+
+    def __resize(self) -> None:
+        new_capacity: int = self.capacity * 2
+        new_buckets: list[None | list[tuple]] = [None] * new_capacity
+        for bucket in self.buckets:
+            if bucket is not None:
+                for key, value in bucket:
+                    new_index: int = hash(key) % new_capacity
+                    if new_buckets[new_index] is None:
+                        new_buckets[new_index] = [(key, value)]
+                    else:
+                        new_buckets[new_index].append((key, value))
+        self.buckets = new_buckets
+        self.capacity = new_capacity
+
+    def __setitem__(self, key: any, value: any) -> None:
+        if self.size / self.capacity >= self.load_factor:
+            self.__resize()
+        index: int = self.__hash_key(key)
+        if self.buckets[index] is None:
+            self.buckets[index] = [(key, value)]
+            self.size += 1
+        else:
+            for i, (k, v) in enumerate(self.buckets[index]):
+                if k == key:
+                    self.buckets[index][i] = (key, value)
+                    return
+            self.buckets[index].append((key, value))
+            self.size += 1
```

Comment:
The `__setitem__` method has a time complexity of O(1) in the best case and O(n) in the worst case due to collision handling, which meets the checklist requirements.